swupd/delta.go

```diff
@@ -42,152 +43,123 @@ func CreateDeltas(manifest, statedir string, from, to uint32) ([]*File, error) {
-		fromToString := fmt.Sprintf("%d-%d-%s-%s", file.DeltaPeer.Version, file.Version, file.DeltaPeer.Hash, file.Hash)
-
-		deltaFile := filepath.Join(deltadir, fromToString)
-
-		// Only add to list if the delta file does not already exist on the system and has enough size.
-		if _, err := os.Stat(deltaFile); os.IsNotExist(err) {
-			fullname := filepath.Join(c.imageBase, fmt.Sprintf("%d", file.Version), "full", file.Name)
-			file.Info, err = os.Stat(fullname)
+		if nf.Version > of.Version && nf.Present() && of.Present() && nf.Type == TypeFile && !sameFile(nf, of) {
+			// Check file size to decide whether it should have a delta.
+			newPath := filepath.Join(c.imageBase, fmt.Sprint(nf.Version), "full", nf.Name)
+			fi, err := os.Stat(newPath)
 			if err != nil {
 				return nil, errors.Wrapf(err, "error accessing %s to decide whether it can have a delta or not")
 			}
-			if file.Info.Size() < minimumSizeToMakeDeltaInBytes {
-				continue
+			if fi.Size() >= minimumSizeToMakeDeltaInBytes {
+				result = append(result, Delta{From: of, To: nf})
 			}
-			files = append(files, file)
 		}
-
-	}
-
-	return files, nil
-}
-
-func createDelta(c config, file *File, statedir string, fromVersion uint32, fromHash Hashval) error {
-	// File without peers cannot have deltas
-	if file.DeltaPeer == nil {
-		return nil
-	}
-
-	// We only support deltas between two regular files for now
-	// TODO: Support directory -> file in the future, this is a complex case
-	if file.Type != TypeFile || file.DeltaPeer.Type != TypeFile {
-		return nil
-	}
-
-	newfile := filepath.Join(c.imageBase, fmt.Sprintf("%d", file.Version), "full", file.Name)
-	original := filepath.Join(c.imageBase, fmt.Sprintf("%d", fromVersion), "full", file.Name)
-
-	deltadir := filepath.Join(c.outputDir, fmt.Sprintf("%d", file.Version), "delta")
-	fromToString := fmt.Sprintf("%d-%d-%s-%s", fromVersion, file.Version, fromHash, file.Hash)
-
-	// Files to create and validate deltas with
-	deltafile := filepath.Join(deltadir, fromToString)
-	testnewfile := filepath.Join(deltadir, "."+fromToString+".testnewfile")
-
-	// Shell out to bsdiff for now...false = don't print to screen
-	if err := helpers.RunCommandSilent("bsdiff", original, newfile, deltafile); err != nil {
-		_ = os.Remove(deltafile) // Might have returned FULLDL
-		return errors.Wrapf(err, "Failed to create delta for %s -> %s", original, newfile)
-	}
-
-	// Check that the delta actually applies correctly
-	if err := helpers.RunCommandSilent("bspatch", original, testnewfile, deltafile); err != nil {
-		return errors.Wrapf(err, "Failed to apply delta %s", deltafile)
+		newIndex++
+		oldIndex++
 	}
 
-	deltahash, err := Hashcalc(testnewfile)
-	if err != nil {
-		return errors.Wrap(err, "Failed to calculate hash for new file")
-	}
+	// TODO: Figure out relationship between deltas and renames. Ideally we want to
+	// call rename detection from here and return more deltas.
```

Comment:
My thought was that the linkPeerAndChange would be doing this walk of the two manifests. If finds all the files which have the same name, produces two lists offiles whose name differs (added and removed), and the rename detection walks these two lists to find things which look as if they are a good match.

See the rename-wip branch for some code which is a couple of weeks old.

Reply:
Tks. I’ll take a look!

Reply:
OK, to avoid conflicts with your work, I'll try to keep the "DeltaPeer" setting around and after that build up my list -- a bit different than what the WIP here was doing, but not so much.